src/cairoUtilFuncGen/abi/abiDecode.ts

```diff
@@ -0,0 +1,179 @@
+import { TranspileFailedError } from '../../utils/errors';
+import { createCairoFunctionStub, createCallToFunction } from '../../utils/functionGeneration';
+import { createBytesTypeName } from '../../utils/nodeTemplates';
+import { getPackedByteSize, isValueType, safeGetNodeType } from '../../utils/nodeTypeProcessing';
+import { typeNameFromTypeNode } from '../../utils/utils';
+import { add, delegateBasedOnType, StringIndexedFuncGenWithAuxiliar } from '../base';
+import { MemoryReadGen } from '../memory/memoryRead';
+
+const IMPLICITS =
+  '{bitwise_ptr : BitwiseBuiltin*, range_check_ptr : felt, warp_memory : DictAccess*}';
+
+export class AbiDecode extends StringIndexedFuncGenWithAuxiliar {
+  protected functionName = 'abi_decode';
+  protected memoryRead: MemoryReadGen;
+
+  constructor(memoryRead: MemoryReadGen, ast: AST, sourceUnit: SourceUnit) {
+    super(ast, sourceUnit);
+    this.memoryRead = memoryRead;
+  }
+
+  public gen(expressions: Expression[], sourceUnit?: SourceUnit): FunctionCall {
+    assert(
+      expressions.length === 2,
+      'ABI decode must recieve two arguments: data to decode, and types to decode into',
+    );
+    const [data, types] = expressions.map(
+      (t) => generalizeType(safeGetNodeType(t, this.ast.compilerVersion))[0],
+    );
+    assert(
+      data instanceof BytesType,
+      `Data must be of BytesType instead of ${printTypeNode(data)}`,
+    );
+    assert(
+      types instanceof TupleType || types instanceof TypeNameType,
+      `Types must be of TupleType or TypeNameType instead of ${printTypeNode(types)}`,
+    );
+    const typesToDecode =
+      types instanceof TupleType
+        ? types.elements.map((t) => {
+            assert(t instanceof TypeNameType);
+            return t.type;
+          })
+        : [types.type];
+
+    const functionName = this.getOrCreate(typesToDecode);
+
+    const functionStub = createCairoFunctionStub(
+      functionName,
+      [['data', createBytesTypeName(this.ast), DataLocation.Memory]],
+      typesToDecode.map((t, index) =>
+        isValueType(t)
+          ? [`result${index}`, typeNameFromTypeNode(t, this.ast)]
+          : [`result${index}`, typeNameFromTypeNode(t, this.ast), DataLocation.Memory],
+      ),
+      ['bitwise_ptr', 'range_check_ptr', 'warp_memory'],
+      this.ast,
+      sourceUnit ?? this.sourceUnit,
+    );
+
+    return createCallToFunction(functionStub, [expressions[0]], this.ast);
+  }
+
+  public getOrCreate(types: TypeNode[]): string {
+    const key = types.map((t) => t.pp()).join(',');
+    const existing = this.generatedFunctions.get(key);
+    if (existing !== undefined) {
+      return existing.name;
+    }
+
+    const [returnParams, encodings] = types.reduce(
+      ([returnParams, encodings], type, index) => {
+        const cairoType = CairoType.fromSol(type, this.ast, TypeConversionContext.Ref);
+        returnParams.push({ name: `result${index}`, type: cairoType.toString() });
```

Comment:
instead of mutating the accumulator, we should use the immutable equivalent
so instead of push we should use concat or spread the original array